### PR TITLE
[FLINK-25229][table] Introduce flink-table-api-bridge-common

### PR DESCRIPTION
**Pull request by slinkydeveloper**
## What is the purpose of the change

Introduce `flink-table-api-bridge-common`. This is required to load the `DataStreamQueryOperation` and `ExternalQueryOperation` without loading scala-api-bridge in planner. This also allowed us to deduplicate `JavaDataStreamQueryOperation` and `ScalaDataStreamQueryOperation`, which are essentially the same class.

Together with it, I also introduced `StreamExecutorFactory` to remove the reflections based hack to use the `ExecutorFactory#create(StreamExecutionEnvironment)` method.

## Brief change log

* Introduce `flink-table-api-bridge-common`
* Deduplicate `DataStreamQueryOperation` and `ExternalQueryOperation` and move to `flink-table-api-bridge-common`
* Introduced `StreamExecutorFactory` and changed call sites which were using reflections
* Introduced `BaseStreamTableEnvironmentImpl` to remove some duplicated code for java and scala `StreamTableEnvironmentImpl`

## Verifying this change

This change is a trivial rework / code cleanup. This is widely covered by e2e tests, examples and planner tests which are using both flink-table-api-scala-bridge and flink-table-api-java-bridge 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-table/flink-table-api-bridge-common/src/main/java/org/apache/flink/table/api/bridge/internal/BaseStreamTableEnvironmentImpl.java [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+import javax.annotation.Nullable;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/** Base utilities to implement a {@code StreamTableEnvironment}. */
+@Internal
+public abstract class BaseStreamTableEnvironmentImpl extends TableEnvironmentImpl {
```
Comment (twalthr) [inline flink-table/flink-table-api-bridge-common/src/main/java/org/apache/flink/table/api/bridge/internal/BaseStreamTableEnvironmentImpl.java:73]: default scope and call it `AbstractStreamTableEnvironmentImpl`

File: flink-connectors/flink-connector-hbase-1.4/pom.xml [hunk 2]
```diff
@@ -133,6 +133,12 @@ under the License.
 			<scope>test</scope>
 		</dependency>
 
+		<dependency>
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-table-api-scala-bridge_${scala.binary.version}</artifactId>
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-connectors/flink-connector-hbase-1.4/pom.xml:138]: Base was also originally planned in FLIP-32:
https://cwiki.apache.org/confluence/display/FLINK/FLIP-32%3A+Restructure+flink-table+for+future+contributions

File: flink-table/flink-table-api-bridge-common/src/main/java/org/apache/flink/table/operations/DataStreamQueryOperation.java [hunk 3]
```diff
@@ -37,30 +37,30 @@
...
  * <p>This operation may expose only part, or change the order of the fields available in a {@link
  * org.apache.flink.api.common.typeutils.CompositeType} of the underlying {@link DataStream}. The
- * {@link ScalaDataStreamQueryOperation#getFieldIndices()} describes the mapping between fields of
- * the {@link TableSchema} to the {@link org.apache.flink.api.common.typeutils.CompositeType}.
+ * {@link DataStreamQueryOperation#getFieldIndices()} describes the mapping between fields of the
+ * {@link TableSchema} to the {@link org.apache.flink.api.common.typeutils.CompositeType}.
  */
 @Internal
-public class ScalaDataStreamQueryOperation<E> implements QueryOperation {
+public class DataStreamQueryOperation<E> implements QueryOperation {
```
Comment (twalthr) [inline flink-table/flink-table-api-bridge-common/src/main/java/org/apache/flink/table/operations/DataStreamQueryOperation.java:44]: feel free to also add `@Deprecated`

File: flink-table/flink-table-api-bridge-common/src/main/java/org/apache/flink/table/api/bridge/internal/BaseStreamTableEnvironmentImpl.java [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+                    identifier,
+                    operation.getDataStream(),
+                    operation.getFieldIndices(),
+                    operation.getResolvedSchema());
+        } else {
+            return queryOperation;
+        }
+    }
+
+    public void attachAsDataStream(List<ModifyOperation> modifyOperations) {
```
Comment (twalthr) [inline flink-table/flink-table-api-bridge-common/src/main/java/org/apache/flink/table/api/bridge/internal/BaseStreamTableEnvironmentImpl.java:325]: the old one was default scoped

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/delegation/ExecutorFactory.java [hunk 5]
```diff
@@ -32,14 +31,9 @@
...
- *
- * <pre>
- *   public Executor create(Configuration, StreamExecutionEnvironment);
- * </pre>
- *
- * <p>This method will be used when instantiating a {@link TableEnvironment} from one of the
- * bridging modules which enables conversion from/to {@code DataStream} API.
+ * <p><b>Important:</b> In order to support DataStream APIs, implementations of this interface must
+ * also implement {@code StreamExecutorFactory} from the {@code flink-table-api-bridge-common}
+ * package.
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/delegation/ExecutorFactory.java:36]: `package` -> `module`

File: flink-table/flink-table-api-scala-bridge/src/main/scala/org/apache/flink/table/api/bridge/scala/internal/StreamTableEnvironmentImpl.scala [hunk 6]
```diff
@@ -326,10 +204,7 @@ class StreamTableEnvironmentImpl (
   override def registerDataStream[T](
       name: String,
       dataStream: DataStream[T],
-      fields: Expression*)
```
Comment (twalthr) [inline flink-table/flink-table-api-scala-bridge/src/main/scala/org/apache/flink/table/api/bridge/scala/internal/StreamTableEnvironmentImpl.scala:329]: this is an unnecessary change and only personal taste, please revert

File: flink-table/flink-table-planner/pom.xml [hunk 7]
```diff
@@ -57,36 +57,12 @@ under the License.
 
 		<!-- core dependencies -->
 
-		<dependency>
-			<groupId>org.apache.flink</groupId>
```
Comment (twalthr) [inline flink-table/flink-table-planner/pom.xml:61]: a pom should declare all dependencies that are directly used, so both `flink-table-common` and `flink-table-api-java`, and `bridge-base` are valid

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/QueryOperationConverter.java [hunk 8]
```diff
@@ -431,21 +428,14 @@ private void splitToProjectionsAndLiterals(
...
-                        externalQueryOperation.getDataStream(),
-                        externalQueryOperation.getPhysicalDataType(),
-                        externalQueryOperation.isTopLevelRecord(),
-                        externalQueryOperation.getChangelogMode(),
-                        externalQueryOperation.getResolvedSchema());
-            } else if (other instanceof ScalaExternalQueryOperation) {
-                final ScalaExternalQueryOperation<?> externalQueryOperation =
-                        (ScalaExternalQueryOperation<?>) other;
+            } else if (other
+                    instanceof org.apache.flink.table.planner.operations.DataStreamQueryOperation) {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/QueryOperationConverter.java:432]: call this `InternalDataStreamQueryOperation` so that we can import it? we should drop it at some point.

**Review by twalthr — CHANGES_REQUESTED** [review]
Very nice clean up @slinkydeveloper. It was definitely worth to do it. I left some minor comments.

File: flink-connectors/flink-connector-hbase-1.4/pom.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-connectors/flink-connector-hbase-1.4/pom.xml:138]: Every user of `test-jar` needs to use api-scala-bridge because of the base tests

File: flink-table/flink-table-api-bridge-common/src/main/java/org/apache/flink/table/operations/DataStreamQueryOperation.java (hunk 3 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-api-bridge-common/src/main/java/org/apache/flink/table/operations/DataStreamQueryOperation.java:44]: Why?

File: flink-table/flink-table-api-bridge-common/src/main/java/org/apache/flink/table/api/bridge/internal/BaseStreamTableEnvironmentImpl.java (hunk 1 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-api-bridge-common/src/main/java/org/apache/flink/table/api/bridge/internal/BaseStreamTableEnvironmentImpl.java:73]: I can't place it in default scope, as the package is `org.apache.flink.table.api.bridge.internal`, while for java is `org.apache.flink.table.api.bridge.java.internal` and for scala `org.apache.flink.table.api.bridge.scala.internal`.

I also can't unify these packages, as the class names will clash.

File: flink-table/flink-table-api-bridge-common/src/main/java/org/apache/flink/table/api/bridge/internal/BaseStreamTableEnvironmentImpl.java (hunk 4 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-api-bridge-common/src/main/java/org/apache/flink/table/api/bridge/internal/BaseStreamTableEnvironmentImpl.java:325]: Same reason as this one https://github.com/apache/flink/pull/18065#discussion_r766516674

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/LegacyDataStreamQueryOperation.java [hunk 9]
```diff
@@ -55,7 +55,7 @@
     private final boolean[] fieldNullables;
     private final FlinkStatistic statistic;
 
-    public DataStreamQueryOperation(
+    public LegacyDataStreamQueryOperation(
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/LegacyDataStreamQueryOperation.java:58]: This one is actually `Internal` and not `Legacy`. It is only used for testing. The other one is `Legacy`.

**Review by twalthr — CHANGES_REQUESTED** [review]
I added one last comment. Should be good to merge afterwards.

**Review by twalthr — APPROVED** [review]
LGTM
